azkeys/coseSigner.go

```diff
@@ -20,32 +20,83 @@ import (
 	"github.com/veraison/go-cose"
 )
 
-// CoseSignerKeyVault is the azure keyvault client for interacting with keyvault keys
-// using a cose.Signer interface
-type CoseSignerKeyVault struct {
-	keyName string
-	alg     keyvault.JSONWebKeySignatureAlgorithm
-	*KeyVault
+// IdentifiableCoseSigner represents a Cose1 signer that has additional methods to provide
+// sufficient information to verify the signed product (an identifier for the signing key and the
+// public key.)
+type IdentifiableCoseSigner interface {
+	cose.Signer
+	PublicKey() (*ecdsa.PublicKey, error)
+	KeyIdentifier() string
+}
+
+// IdentifiableCoseSignerFactory is for creating IdentifiableCoseSigners. The reason for a factory
+// here is that we can always create a fresh instance, capturing the latest key information at that
+// point in time.
+type IdentifiableCoseSignerFactory interface {
+	NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error)
 }
 
-// NewCoseSignerKeyVault creates a new keyvault configuration that signs with ES384
+// KeyVaultCoseSignerFactory creates instances of our Azure KeyVault implementation of
+// IdentifiableCoseSigner. The keyvault configuration is stored on the object and new instances
+// can be created without caller knowledge of it.
+type KeyVaultCoseSignerFactory struct {
+	keyVaultURL string
+	keyName     string
+}
+
+// NewKeyVaultCoseSignerFactory returns a new instance of the factory, storing the keyvault config
+func NewKeyVaultCoseSignerFactory(keyVaultURL string, keyName string) *KeyVaultCoseSignerFactory {
+	return &KeyVaultCoseSignerFactory{
+		keyVaultURL: keyVaultURL,
+		keyName:     keyName,
+	}
+}
+
+// NewIdentifiableCoseSigner creates a new keyvault configuration that signs with ES384
 // using the latest version of the named key
-func NewCoseSignerKeyVault(keyVaultURL string, keyName string) *CoseSignerKeyVault {
-	kv := CoseSignerKeyVault{
-		keyName: keyName,
+func (f *KeyVaultCoseSignerFactory) NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error) {
+	kv := KeyVaultCoseSigner{
+		keyName: f.keyName,
 		alg:     keyvault.ES384, // hardwired for the moment, add caller setting when needed
 		KeyVault: &KeyVault{
-			url: keyVaultURL,
+			url: f.keyVaultURL,
 		},
 	}
-	return &kv
+
```

Comment:
Good choice in not logging in this level. Its the callers job to do that I think